### PR TITLE
[BCEL-317] Pluggable cache for ConstantUtf8

### PR DESCRIPTION
**Pull request by suztomo**
https://issues.apache.org/jira/browse/BCEL-317

- `getCachedInstance` is unused in BCEL.
- The caller of `getInstance` is a static method `Constant.readConstant`.

image

### CODE REVIEW

File: src/main/java/org/apache/bcel/classfile/ConstantUtf8.java [hunk 1]
```diff
@@ -88,39 +128,11 @@ public void run() {
         }
     }
 
-    /**
```
Comment (garydgregory) [inline src/main/java/org/apache/bcel/classfile/ConstantUtf8.java]: You cannot remove existing public methods. That causes a break in binary compatibility.

Comment (suztomo) [reply src/main/java/org/apache/bcel/classfile/ConstantUtf8.java]: Let me fix that

**Comment by garydgregory** [comment]
I'm not sure if this PR is a good thing. I'd like feedback from the community.

**TL;DR**

If you want caching, use the existing `getCachedInstance()` API, it's an existing feature.

**Details**
[Please do not hesitate to correct details I may have wrong]

Right now, before this PR, we have two clear and simple APIs:
- `getInstance()` always returns a new instance.
- `getCachedInstance()` returns a cached instance unless the cache is full.

Some thoughts:
- The PR adds 2 global variables as system properties, this is definitely not a good thing, even if the current code already uses a sys prop.
- The PR removes an existing sys prop, therefore breaking compatibility.
- The PR changes the behavior of `getInstance()` which can now return a cached instance, which could (or not) break things but definitely falls into the unexpected category IMO.

**More Details** 

The PR should not write to the console.

The existing code might be better re-implemented using a ConcurrentHashMap for simpler synchronization.

**Comment by suztomo** [comment]
@garydgregory Thank you for the feedback. My approach in this PR tries to avoid modifying other classes than `ConstantUtf8`. 

While waiting for other maintainers' comments, I will think about another approach that does not change the behavior of `getInstance` and `getCachedInstance`.

Removed `System.out.println` from this PR.

**Comment by garydgregory** [comment]
Hi @suztomo 

I can appreciate the desire to provide a PR with a narrow footprint, but we must take semantics into account here. My basic issue is still that we have two public APIs, one that says "give me an instance" and the other that says "give me a cached instance". This to me implies that the former never returns a cached instance and it currently does not indeed. The latter always returns a cached instance, but in fact only tries to do so by enforcing a cache size limit. 

It feels like this PR is trying to force caching on clients that use getInstance() instead of properly calling getCachedInstance(). If the client is BCEL itself, then we need to address using BCEL in a caching vs. not mode.

**Comment by suztomo** [comment]
I had an (wrong) impression that `getInstance` can return a cached instance. It's because until the performance degradation problem ([BCEL-186](https://issues.apache.org/jira/browse/BCEL-186)), `getInstance` used to return a cached result depending on the system property ([code diff to revert that 4 years ago](https://github.com/apache/commons-bcel/commit/fbcd823d35677e12d3bc37db531f3f18cdb17535#diff-caf2fe667754dbaaa32f5e4a9bfe67e4L93
)). This PR is on top of the reverted design but not caching by default.

**Comment by suztomo** [comment]
Closing this in favor of #27
